Waterway overhaul

tburrows13 opened the pull request:
Replaces waterway free recipe with a shortcut.

Ghost waterways are now immediately 'revived', and marked-for-deconstruction waterways are now immediately destroyed.

Refactored waterway and landfill code to use vanilla collision masks rather than control scripting.

tburrows13 committed:
Remove water-way-placed, prevent waterways on land by adding new collision layer to ground tiles

tburrows13 committed:
Waterways revive on ghost-created and destroy on marked-for-deconstruction

tburrows13 committed:
Remove landfill scripting, use check_collision_with_entities

tburrows13 committed:
Add waterway shortcut, remove recipe

tburrows13 committed:
Refactory CheckRailPlacement

robot256 commented on data-final-fixes.lua:
```diff
@@ -1,8 +1,19 @@
+collision_mask_util.add_layer(data.raw["straight-rail"]["straight-water-way"].collision_mask, waterway_layer)
+collision_mask_util.add_layer(data.raw["curved-rail"]["curved-water-way"].collision_mask, waterway_layer)
+collision_mask_util.add_layer(data.raw["straight-rail"]["invisible_rail"].collision_mask, waterway_layer)
+collision_mask_util.add_layer(data.raw["straight-rail"]["bridge_crossing"].collision_mask, waterway_layer)
```
Need to check that this doesn't interfere with bridge placement.  The land train tracks might get weird if it's on a narrow water channel.

tburrows13 replied:
I've tested it and it seems to work even on narrow channels like this.
![Screenshot 2023-03-29 at 00 22 51](https://user-images.githubusercontent.com/22625968/228388574-206d4fbc-3043-4a11-ba0a-b5e7e54534f5.png)

robot256 replied:
What about placing the bridge in water and landfilling under the land tracks?  When the bridge is first placed, the invisible rails are built by script which doesn't check collision masks.

tburrows13 replied:
All seems to work fine. In order:
![Screenshot 2023-03-30 at 23 48 02](https://user-images.githubusercontent.com/22625968/228981995-4e1b0272-296f-4a84-8547-bd728e1edbd7.png)
![Screenshot 2023-03-30 at 23 48 11](https://user-images.githubusercontent.com/22625968/228981999-ca75f1a0-a0ba-48cd-9b5d-add8bc38a5d8.png)
![Screenshot 2023-03-30 at 23 49 21](https://user-images.githubusercontent.com/22625968/228982005-5e7ec716-4dea-4232-8ca7-3827cbb9907b.png)

robot256 commented on prototypes/water_rails.lua:
```diff
@@ -136,17 +136,16 @@ data:extend({
+    flags = {"placeable-neutral", "player-creation", "building-direction-8-way"},
     resistances = invincible,
-    minable = {mining_time = 0.2, result = "water-way"},
+    minable = {mining_time = 0.1},
```
Only reason for a longer mining time is to prevent accidental clicks from mining a hole in your rail network.  If you need to do a lot of them the deconstruction planner will be faster.

tburrows13 committed:
Revert mining_time changes

tburrows13 committed:
Revive all ghost water-ways in migration

tburrows13 committed:
Remove no-longer-used event handlers

tburrows13 committed:
Rename water-way item so that all previous water-way items are removed on load

tburrows13 committed:
Prevent playing hundreds of revive sounds simultaneously

tburrows13 committed:
Add waterway shortcut icon

tburrows13 commented:
<img width="373" alt="231596883-6e6fb718-ff1a-43e9-abba-88c8363b0bfd" src="https://user-images.githubusercontent.com/22625968/231597776-34b096d3-7724-4cfd-9a07-25a3dd9cfeb3.png">

Required attribution: <a target="_blank" href="https://icons8.com/icon/66157/fishing-boat">Fishing Boat</a> icon by <a target="_blank" href="https://icons8.com">Icons8</a>

tburrows13 committed:
Rename waterway item again to a simpler name

tburrows13 committed:
"Water way" -> "Waterway", various locale cleanup
Generally better not to use __ITEM__ and __ENTITY__ tags mid-sentence because it adds capital letters for no reason.

tburrows13 committed:
Locale cleanup, remove many mid-sentence capitals

tburrows13 committed:
Typo fix

tburrows13 committed:
Merge branch 'hotfix2' into waterway-overhaul

tburrows13 committed:
Merge branch 'hotfix2' into waterway-overhaul

tburrows13 committed:
Fix broken merge

tburrows13 committed:
Add waterway default keybind and disabled icon
